plutus-metatheory/src/Main.lagda

```diff
@@ -262,21 +262,28 @@ reportError (runtimeError userError)        = "userError"
 reportError (runtimeError runtimeTypeError) = "runtimeTypeError"
 
 
+{- check if the term is an error term, and in that case 
+  return an ERROR. 
+  This is used when evaluation of the reduction semantics has ended
+-}
+checkError : ∀{A} → ∅ ⊢ A → Either ERROR (∅ ⊢ A )
+checkError (error _) = inj₁ (runtimeError userError)
+checkError t         = return t
+
 executePLC : EvalMode → ScopedTm Z → Either ERROR String
 executePLC U t = do
   (A ,, t) ← withE (λ e → typeError (uglyTypeError e)) $ typeCheckPLC t
   □ V ← withE runtimeError $ U.stepper maxsteps (ε ; [] ▻ erase t)
     where ◆  → inj₁ (runtimeError userError)
           _    → inj₁ (runtimeError gasError)
-
 {-
 just t' ← withE runtimeError $ U.stepper maxsteps (ε ; [] ▻ erase t)
     where nothing → inj₁ (runtimeError userError)
     -}
   return $ prettyPrintUTm (U.extricateU0 (U.discharge V))
 executePLC TL t = do
   (A ,, t) ← withE (λ e → typeError (uglyTypeError e)) $ typeCheckPLC t
-  t' ← withE runtimeError $ L.stepper t maxsteps
+  t' ← (withE runtimeError $ L.stepper t maxsteps) >>= checkError
```

Comment:
I'm a bit confused, I thought `withE` was exactly supposed to say what to do when you hit an error term? So how can we end up in the situation where we just return an error term?

Reply:
`withE` just handles the `Left error` case.  If you look at the type it doesn't even know about Terms.
```
withE : {A B C : Set} → (A → B) → Either A C → Either B C
```
The reduction semantics (`L.stepper`) never raises an error when it encounters an error term, it just returns `Right (error term)`:

```
... | steps x (error _)   = return (error A)
```
 My first instinct was to change `L.stepper` to raise an error but that broke the NEAT tests.

Reply:
Ah okay, then this makes perfect sense!